### PR TITLE
Make RAND_DRBG fork-safe

### PR DESCRIPTION
**Pull request by richsalz**
Use atfork to count child forks, and reseed DRBG when the counts don't
match.

### CODE REVIEW

**Comment by kroeckx** [comment]
Do we need to acquire locks in the before function and release them in the parent and child?

**Comment by richsalz** [comment]
I don't know; do we?

**Comment by richsalz** [comment]
I don't think we need to do this.  The current code only bumps a global variable in the child.  I don't think any locking is necessary -- it's handled by the existing locks and API.

File: crypto/rand/rand_lcl.h [hunk 1]
```diff
@@ -127,6 +128,9 @@ extern RAND_BYTES_BUFFER rand_bytes;
 extern RAND_DRBG rand_drbg;
 extern RAND_DRBG priv_drbg;
 
+/* How often we've fork (only incremented in child). */
```
Comment (paulidale) [inline crypto/rand/rand_lcl.h]: s/fork/forked/ or put _called_ before fork ?

File: crypto/rand/drbg_lib.c [hunk 2]
```diff
@@ -72,6 +72,7 @@ RAND_DRBG *RAND_DRBG_new(int type, unsigned int flags, RAND_DRBG *parent)
     }
     drbg->size = RANDOMNESS_NEEDED;
     drbg->randomness = ucp;
+    drbg->fork_count = rand_fork_count;
```
Comment (dot-asm) [inline crypto/rand/drbg_lib.c]: While you are at it, wouldn't it be appropriate to move this empty line below?

Comment (richsalz) [inline crypto/rand/drbg_lib.c]: I just removed the line and left a single block of init statements/functions.  thanks.  assuming the +1 still holds.

**Comment by richsalz** [comment]
I fixed the comment nit, merged, all tests pass, closing.  thanks.

File: CHANGES [hunk 3]
```diff
@@ -9,6 +9,11 @@
 
  Changes between 1.1.0f and 1.1.1 [xx XXX xxxx]
 
+  *) Add "atfork" functions.  If building on a system that without
+     pthreads, see doc/man3/OPENSSL_fork_prepare.pod for application
+     requirements.  The RAND facility now uses/requires this.
```
Comment (dot-asm) [inline CHANGES]: There is ambiguity in this statement. It suggests that *all* systems have fork() and some of them don't have pthreads. There are systems/environments that have pthreads, but don't actually have fork. But more importantly try to find fork-capable yet pthreads-incapable system...

Comment (richsalz) [inline CHANGES]: I could not think of a better way to write it.  The manpage is very explicit however.

**Pull request closed without merging**